jep/302/README.adoc

```diff
@@ -0,0 +1,260 @@
+
+_Workspace Root Directory_:: `${JENKINS_HOME}/workspace/${ITEM_FULLNAME}`
+_Build Record Root Directory_::
+`${JENKINS_HOME}/var/${ITEM_FULL_NAME}/builds`
+
+Though we will for now backup these directories too, having this distinction baked in from the ground up could prove useful in the future if we need to make the snapshotting and rollback processes smarter and more robust.
+
+For instance, in a case where Jenkins would not come back online after an automated rollback, we would be able to leverage this data separation to try a degraded rollback strategy where we would handle that `$JENKINS_HOME/var` directory in a different way than `JENKINS_HOME/jobs` or other ones.
+
+== Motivation
+
+Jenkins has never supported downgrading by itself, and it's unlikely the core constructs will change in this regard anytime soon.
+The official way to revert an upgrade if something went wrong is to restore a previous backup.
+
+In the context of _Essentials_, we cannot rely on external backups to revert to the _N-1_ version: this would require some manual user intervention, which is clearly not the user experience _Essentials_ wants to provide.
+
+== Reasoning
+
+=== Why Git
+
+Using filesystem-level tools offering a snapshotting feature, like LVM, ZFS or btrfs to give a few examples, was considered.
+But this was evacuated because _Essentials_ vision is about providing an link:https://github.com/jenkinsci/jep/tree/71d9391744c8cc7d6595805f7fdd327eedf6811a/jep/300#automatically-updated-distribution["_easier to use_ and _easier to manage_ Jenkins environment"].
+As per the link:https://github.com/jenkinsci/jep/tree/71d9391744c8cc7d6595805f7fdd327eedf6811a/jep/300#target-audience[targeted audience], we obviously do not want to expect _Essentials_ users to be system experts able to set up a dedicated filesystem to operate Jenkins.
+And even with system expert, doing so would not make Essentials a very easy and quick to use distribution of Jenkins.
+
+Git offers in this matter a powerful user-space tool that allows us to version,
+and quickly rollback to some previous state if need be.
+
+Git is also a very common tool nowadays for developers,
+hence it will help making Essentials more accessible to contributors.
+
+== Why not use compatibleSinceVersion metadata
+
+TIP: A given plugin can indicate what is "the oldest version [...] configuration-compatible with.".
+In other words, if I am changing the configuration format of a plugin in non backwards compatible manner, I am _expected_ to express it as such using link:https://jenkinsci.github.io/maven-hpi-plugin/hpi-mojo.html#compatibleSinceVersion[`compatibleSinceVersion`].
+
+So, this means we could use this field to _optimize_ the way we roll back: once a configuration file has been upgraded by a plugin update, we can exclude it from rollback if the `compatibleSinceVersion` field was not changed.
```

Comment:
Not following how you got there.

What `compatibleSinceVersion` means is this: if this flag is marked with a value of, say, `1.5`, and you have just updated the plugin from version 1.4 to 1.5, but you wish to revert back to 1.4 after seeing some problems, and you can detect that there is an XML file written by that plugin (visible via the `plugin` attribute):

* You _may_ revert that file to the previous version. By design, Essentials always will (modulo the discussion about `build.xml` and similar build record files).
* You _must not_ leave it at the version written by 1.5. So if you are invested in the modifications to this file made recently, you are in trouble.

But you seem to be talking about the opposite: the usual case that the upgrade is _not_ crossing a `compatibleSinceVersion` boundary. In that case it is fine to let the 1.4 plugin read a modification to the config file. But that is irrelevant for this JEP since you are versioning all of `$JENKINS_HOME` just in case.

Reply:
I think my phrasing might have been a bit convoluted, but I think we are saying the same thing.

I wrote, chosen excerpt:
> exclude it from rollback if the `compatibleSinceVersion` field was not changed

I.e. taking your example to illustrate: when going from version `1.4` to `1.5`, *but* with `compatibleSinceVersion` staying at `1.4` or lower, then we _could_ avoid rolling the generated config files for this plugin, even if config files with `... plugin="some-feature@1.5"` present.

Reply:
Heavily rephrased in last commit.